File: src/main/java/frc/robot/commands/intake/IntakeMove.java
package frc.robot.commands.intake;

import edu.wpi.first.wpilibj2.command.CommandBase;
import edu.wpi.first.wpilibj2.command.button.Trigger;
import frc.robot.Constants;
import frc.robot.Robot;
import frc.robot.subsystems.Hopper;
import frc.robot.subsystems.Intake;
import frc.robot.util.TimedTrigger;

/**
 * Use WhileHeld
 */
public class IntakeMove extends CommandBase {

  private double mSpeed = 0.0;
  private boolean mEnableAutoLoader = false;
  private TimedTrigger mIntakeTrigger;

  public IntakeMove(double speed, boolean enableAutoLoader) {
    mSpeed = speed;
    mEnableAutoLoader = enableAutoLoader;

    mIntakeTrigger = new TimedTrigger(Constants.Values.INTAKE_EXTEND_DELAY);

    addRequirements(Intake.getInstance());
  }

  @Override
  public void initialize() {
    if (mEnableAutoLoader)
      Robot.autoLoadHopper = true;

    mIntakeTrigger.trigger();
    Intake.getInstance().setPiston(true);
  }

  @Override
  public void execute() {
    if (mIntakeTrigger.get(false)) {
      if (mEnableAutoLoader) {
        if (!Hopper.getInstance().getShooterBall())
          Intake.getInstance().setPercent(mSpeed);
        else
          Intake.getInstance().setPercent(0.0);
      } else {
        Intake.getInstance().setPercent(mSpeed);
      }
    }
  }

  @Override
  public void end(boolean interrupted) {
    Intake.getInstance().setPercent(0.0);
    Robot.autoLoadHopper = false;
    Intake.getInstance().setPiston(false);
  }

}

File: src/main/kotlin/frc/robot/commands/intake/IntakeMove.kt
package frc.robot.commands.intake

import edu.wpi.first.wpilibj2.command.CommandBase
import frc.robot.Constants
import frc.robot.Robot
import frc.robot.subsystems.Hopper
import frc.robot.subsystems.Intake
import frc.robot.utils.TimedTrigger

class IntakeMove(val speed: Double, val enableAutoLoader: Boolean): CommandBase() {

  private val mPistonTrigger: TimedTrigger = TimedTrigger(Constants.Values.INTAKE_EXTEND_DELAY)

  init {
    addRequirements(Intake)
  }

  override fun initialize() {
    Intake.setPiston(true)
    mPistonTrigger.reset()
    mPistonTrigger.trigger()
  }

  override fun execute() {
    if (mPistonTrigger.get(false)) {
      Robot.EnableAutoLoader = enableAutoLoader
      if (enableAutoLoader) {
        if (!Hopper.getShooterBall()) Intake.setMotor(speed) // TODO: Get shooter ball
        else Intake.setMotor(0.0)
      } else { // Basically only for ejecting balls
        Intake.setMotor(speed)
      }
    } else {
      Intake.setMotor(0.0)
    }
  }

  override fun end(interrupted: Boolean) {
    Intake.setMotor(0.0)
    Robot.EnableAutoLoader = false;
    Intake.setPiston(false)
  }

}